branching_xblock/branching_xblock.py

```diff
@@ -81,6 +83,9 @@ class BranchingXBlock(XBlock):
         help="Accumulated score (if scoring enabled)"
     )
 
+    # Do we need to store this since it canbe calculated at runtime?
+    # There could be issues if someone gets this marked as completed and then
+    # more choices are added.
```

Comment:
That's a good point. We can calculate it at runtime but we would still need to trigger the completion and grade submission when the user first reaches a leaf node. Any suggestions on how to handle that?

Reply:
I think the way you're handling it currently is perfectly fine, it's just one value that won't be set. 

I see the following places where `has_completed` is use:
- Returned as part of `_get_state`
- Set to true when user reaches a leaf/end node
- Checked along with `enable_scoring` to reset score to 0 to publish grade
- Set to false on undo

I think in all these cases, you can simply replace with a function called `has_completed`. Then you don't need to keep updating it, it will automatically be derived. 

I also don't see it used in the frontend, so it need not be part of the API state.